atef/qt_helpers.py

```diff
@@ -271,3 +276,53 @@ def put_to_index(self, index: int, new_value: Any) -> None:
         self.changed_value.emit(new_value)
         self.changed_index.emit(index)
         self.updated.emit()
+
+
+class ThreadWorker(QtCore.QThread):
+    """
+    Worker thread helper.  For running a function in a background QThread.
+
+    Parameters
+    ----------
+    func : callable
+        The function to call when the thread starts.
+    *args
+        Arguments for the function call.
+    **kwargs
+        Keyword arguments for the function call.
+    """
+
+    error_raised = QtCore.Signal(Exception)
+    returned = QtCore.Signal(object)
+    func: Callable
+    args: Tuple[Any, ...]
+    kwargs: Dict[str, Any]
+    return_value: Any
+
+    def __init__(self, func: Callable, *args, **kwargs):
+        super().__init__()
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+        self.return_value = None
+
+    @QtCore.Slot()
+    def run(self):
+        try:
+            self.return_value = self.func(*self.args, **self.kwargs)
+        except Exception as ex:
+            logger.exception(
+                "Failed to run %s(*%s, **%r) in thread pool",
+                self.func,
+                self.args,
+                self.kwargs,
+            )
+            self.return_value = ex
+            self.error_raised.emit(ex)
+        else:
+            self.returned.emit(self.return_value)
+
+
+def run_in_gui_thread(func: Callable, *args, _start_delay_ms: int = 0, **kwargs):
```

Comment:
I like this one too. I've occasionally been known to throw out a single shot timer for various reasons and this is a good shortcut.